geode-redis/src/main/java/org/apache/geode/redis/internal/executor/set/RedisSet.java

```diff
@@ -304,70 +211,60 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
   }
 
   /**
-   * @param membersToAdd members to add to this set; NOTE this list may by
-   *        modified by this call
-   * @param region the region this instance is stored in
    * @param key the name of the set to add to
+   * @param membersToAdd members to add to this set; NOTE this list may by modified by this call
    * @return the number of members actually added; -1 if concurrent modification
    */
-  private synchronized long doSadd(ArrayList<ByteArrayWrapper> membersToAdd,
-      Region<ByteArrayWrapper, RedisSet> region,
-      ByteArrayWrapper key) {
-
+  @Override
+  public synchronized long sadd(ByteArrayWrapper key, ArrayList<ByteArrayWrapper> membersToAdd) {
+    this.deltas = null;
     membersToAdd.removeIf(memberToAdd -> !members.add(memberToAdd));
     int membersAdded = membersToAdd.size();
     if (membersAdded != 0) {
       deltasAreAdds = true;
       deltas = membersToAdd;
-      try {
-        region.put(key, this);
-      } finally {
-        deltas = null;
-      }
     }
     return membersAdded;
   }
 
   /**
-   * @param membersToRemove members to remove from this set; NOTE this list may by
-   *        modified by this call
-   * @param region the region this instance is stored in
    * @param key the name of the set to remove from
+   * @param membersToRemove members to remove from this set; NOTE this list may by modified by this
+   *        call
    * @param setWasDeleted set to true if this method deletes the set
    * @return the number of members actually removed; -1 if concurrent modification
    */
-  private synchronized long doSrem(ArrayList<ByteArrayWrapper> membersToRemove,
-      Region<ByteArrayWrapper, RedisSet> region,
-      ByteArrayWrapper key, AtomicBoolean setWasDeleted) {
-
+  @Override
+  public synchronized long srem(ByteArrayWrapper key, ArrayList<ByteArrayWrapper> membersToRemove,
+      AtomicBoolean setWasDeleted) {
+    deltas = null;
```

Comment:
I don't think they way you have refactored the delta code is thread safe. As long as we have set ops that do not use the function+delta (that is why this method is synchronized still) then we need to send the delta before this synchronization is released.
I also don't like that we leave deltas set until the next operation is done. This will just extend its life when it should have been garbage before this method released in sync. I think this issue with the delta management is throughout this class.

Reply:
Maybe we could pass the RedisSetInRegion into the methods that need to update the region. If we added the "key" to RedisSetInRegion then it would know what key it is supposed to remove or put when we callback to it.
This would also allow the RedisSetInRegion and RedisSet to both implement a common interface (neither of their methods would have "key" parameters).

Reply:
ideally, i think we can remove the key from the `RedisSetCommands` interface.  I think that's a more OO approach where the key is the identity of the object and not something you can pass in.  Right now the `RedisSetCommands` interface is more like a bag of functions.  What do you think?  

I think that change is a little bit down the road, though, once all the set commands are   moved over.

Reply:
When JohnH and I worked on renaming RedisSetCommands (from RedisSet), we chose "Commands" because it seemed like a bag of functions. And from the point of view of the netty executors (like SAddExecutor) I think this may make sense. I like for the interface methods to have a close correspondence to the actual redis command line arguments. Since the REDIS SADD takes a "key" and "members" then the method on the Commands interface should have both a "key" and "members". It should not have "region" since the REDIS SADD command does not have a region. So that is why when you create a RedisSetCommandsFunctionExecutor you give it a Region but you do not give it the key. This also means you could use the same instance of RedisSetCommandsFunctionExecutor to do multiple commands on different keys.
But once you get down to having on instance of our data structure (RedisSet in this example) then you no longer need the key because the key was used to find it.
If I had to choose one or the other (methods with the key or methods without), I think I'd chose with the key since it corresponds to the redis apis.

Reply:
I can see both sides of this coin.  

I think it's nice that the `RedisSetCommands` match the Redis API. 
I also think that once we are in the executor we are in the Java code, we are now doing OO programming and don't need our interfaces to conform to Redis' binary protocol.

The "user" I imagine implementing this interface is a developer looking to implement the Redis data structures in a completely new way...in geode or in some other platform.  I don't have enough of that architecture worked out to be able to make a clear decision, so at this point I am okay either way.